File: apps/mobile/src/services/contextHeuristics.ts
export type MemoryNodeType = "entry" | "goal" | "topic" | "mood" | "anchor";

export interface ModeSample {
  mood: string | null;
  created_at: string;
}

export interface MemoryMatchInput {
  node_id: string;
  score: number;
  node_type: MemoryNodeType;
  text: string;
  trust_weight: number;
  sentiment: number | null;
}

export interface MemoryNodeInput {
  id: string;
  type: MemoryNodeType;
  text: string;
  trust_weight: number;
  sentiment: number | null;
}

export interface MemoryEdgeInput {
  id?: string;
  src_id: string;
  dst_id: string;
  relation: string;
  weight: number;
}

export interface EvidenceNode {
  id: string;
  type: MemoryNodeType;
  text: string;
  strength: number;
  trust_weight: number;
  sentiment: number | null;
  sources: string[];
}

export function inferLikelyNeed(
  inputText: string,
  dominantMood: string | null,
  topTopic: string | null
): string {
  const normalized = inputText.toLowerCase();
  if (normalized.includes("schedule") || normalized.includes("calendar")) {
    return "schedule_alignment";
  }
  if (normalized.includes("goal") || normalized.includes("progress")) {
    return "goal_support";
  }
  if (
    normalized.includes("tired") ||
    normalized.includes("fatigue") ||
    normalized.includes("exhausted") ||
    normalized.includes("burned out")
  ) {
    return "energy_check";
  }
  if (normalized.includes("again") && dominantMood === "stressed") {
    return "stress_recovery";
  }
  if (dominantMood === "tired") {
    return "restoration_prompt";
  }
  if (topTopic === "work" || topTopic === "project") {
    return "workload_review";
  }
  if (topTopic === "family" || topTopic === "relationships") {
    return "relationship_context";
  }
  return "context_refresh";
}

export function aggregateModes(entries: ModeSample[]) {
  const counts = new Map<string, { count: number; lastSeen: string }>();
  for (const entry of entries) {
    const mood = entry.mood?.trim().toLowerCase();
    if (!mood) continue;
    const current = counts.get(mood);
    if (current) {
      current.count += 1;
      if (current.lastSeen < entry.created_at) {
        current.lastSeen = entry.created_at;
      }
    } else {
      counts.set(mood, { count: 1, lastSeen: entry.created_at });
    }
  }

  return Array.from(counts.entries())
    .map(([label, value]) => ({
      label,
      count: value.count,
      last_seen_at: value.lastSeen,
    }))
    .sort((a, b) => b.count - a.count || b.last_seen_at.localeCompare(a.last_seen_at))
    .slice(0, 6);
}

export function buildEvidence(
  matches: MemoryMatchInput[],
  edgeRows: MemoryEdgeInput[],
  neighborNodes: Map<string, MemoryNodeInput>,
  includeFatigue: MemoryNodeInput[],
  threshold = 0.4
): EvidenceNode[] {
  const evidence = new Map<
    string,
    {
      id: string;
      type: MemoryNodeType;
      text: string;
      strength: number;
      trust_weight: number;
      sentiment: number | null;
      sources: Set<string>;
    }
  >();

  const addEvidence = (
    node: MemoryNodeInput,
    strength: number,
    source: string
  ) => {
    if (!node || strength < threshold) return;
    const existing = evidence.get(node.id);
    if (existing) {
      if (existing.strength < strength) {
        existing.strength = strength;
      }
      existing.sources.add(source);
    } else {
      evidence.set(node.id, {
        id: node.id,
        type: node.type,
        text: node.text,
        strength,
        trust_weight: node.trust_weight,
        sentiment: node.sentiment,
        sources: new Set([source]),
      });
    }
  };

  for (const match of matches) {
    addEvidence(
      {
        id: match.node_id,
        type: match.node_type,
        text: match.text,
        trust_weight: match.trust_weight,
        sentiment: match.sentiment,
      },
      match.score * match.trust_weight,
      "vector_match"
    );
  }

  for (const edge of edgeRows) {
    const source = neighborNodes.get(edge.src_id);
    const target = neighborNodes.get(edge.dst_id);
    if (!source || !target) continue;

    const forwardStrength = edge.weight * target.trust_weight;
    addEvidence(target, forwardStrength, `edge:${edge.relation}`);

    const reverseStrength = edge.weight * source.trust_weight;
    addEvidence(source, reverseStrength, `edge:${edge.relation}`);
  }

  for (const node of includeFatigue) {
    addEvidence(node, Math.min(0.9, node.trust_weight), "fatigue_recall");
  }

  return Array.from(evidence.values())
    .map((item) => ({
      id: item.id,
      type: item.type,
      text: item.text,
      strength: Number(item.strength.toFixed(3)),
      trust_weight: Number(item.trust_weight.toFixed(3)),
      sentiment: item.sentiment,
      sources: Array.from(item.sources).slice(0, 5),
    }))
    .sort((a, b) => b.strength - a.strength)
    .slice(0, 20);
}

File: services/python-ai/app/orchestrator/context_heuristics.py
from typing import Dict, List, Optional


MemoryNodeType = str


def infer_likely_need(input_text: str, dominant_mood: Optional[str], top_topic: Optional[str]) -> str:
    normalized = input_text.lower()
    if "schedule" in normalized or "calendar" in normalized:
        return "schedule_alignment"
    if "goal" in normalized or "progress" in normalized:
        return "goal_support"
    if any(token in normalized for token in ["tired", "fatigue", "exhausted", "burned out"]):
        return "energy_check"
    if "again" in normalized and dominant_mood == "stressed":
        return "stress_recovery"
    if dominant_mood == "tired":
        return "restoration_prompt"
    if top_topic in ("work", "project"):
        return "workload_review"
    if top_topic in ("family", "relationships"):
        return "relationship_context"
    return "context_refresh"


def aggregate_modes(entries: List[Dict[str, str]]) -> List[Dict[str, str]]:
    counts: Dict[str, Dict[str, any]] = {}
    for entry in entries:
        mood = (entry.get("mood") or "").strip().lower()
        if not mood:
            continue
        current = counts.get(mood)
        if current:
            current["count"] += 1
            if current["lastSeen"] < entry.get("created_at", ""):
                current["lastSeen"] = entry.get("created_at", "")
        else:
            counts[mood] = {"count": 1, "lastSeen": entry.get("created_at", "")}
    result = [
        {"label": label, "count": value["count"], "last_seen_at": value["lastSeen"]}
        for label, value in counts.items()
    ]
    return sorted(result, key=lambda item: (-item["count"], item["last_seen_at"]))[:6]


def build_evidence(matches: List[Dict[str, any]], edge_rows: List[Dict[str, any]], neighbor_nodes: Dict[str, Dict[str, any]], include_fatigue: List[Dict[str, any]], threshold: float = 0.4) -> List[Dict[str, any]]:
    evidence: Dict[str, Dict[str, any]] = {}

    def add_evidence(node: Dict[str, any], strength: float, source: str):
        if not node or strength < threshold:
            return
        existing = evidence.get(node["id"])
        if existing:
            if existing["strength"] < strength:
                existing["strength"] = strength
            existing["sources"].add(source)
        else:
            evidence[node["id"]] = {
                "id": node["id"],
                "type": node.get("type"),
                "text": node.get("text"),
                "strength": strength,
                "trust_weight": node.get("trust_weight", 0),
                "sentiment": node.get("sentiment"),
                "sources": {source},
            }

    for match in matches:
        node = {
            "id": match.get("node_id"),
            "type": match.get("node_type"),
            "text": match.get("text"),
            "trust_weight": match.get("trust_weight", 0),
            "sentiment": match.get("sentiment"),
        }
        add_evidence(node, match.get("score", 0) * match.get("trust_weight", 0), "vector_match")

    for edge in edge_rows:
        source = neighbor_nodes.get(edge.get("src_id"))
        target = neighbor_nodes.get(edge.get("dst_id"))
        if not source or not target:
            continue
        forward_strength = edge.get("weight", 0) * target.get("trust_weight", 0)
        add_evidence(target, forward_strength, f"edge:{edge.get('relation')}")
        reverse_strength = edge.get("weight", 0) * source.get("trust_weight", 0)
        add_evidence(source, reverse_strength, f"edge:{edge.get('relation')}")

    for node in include_fatigue:
        add_evidence(node, min(0.9, node.get("trust_weight", 0)), "fatigue_recall")

    formatted = []
    for item in evidence.values():
        formatted.append(
            {
                "id": item["id"],
                "type": item["type"],
                "text": item["text"],
                "strength": round(item["strength"], 3),
                "trust_weight": round(item["trust_weight"], 3),
                "sentiment": item["sentiment"],
                "sources": list(item["sources"])[:5],
            }
        )
    return sorted(formatted, key=lambda node: node["strength"], reverse=True)[:20]

